### PR TITLE
Added serialization

### PR DESCRIPTION
**Pull request by villanellex2**

### CODE REVIEW

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+          "java.lang.Short",
+          "java.lang.Integer",
+          "java.lang.Long",
+          "java.lang.Float",
+          "java.lang.Double",
+          "java.lang.Boolean",
+          "java.lang.Character",
+          "java.lang.String");
+
+  public void createXMLDocument(Stream<Object> inputStream, StreamResult stream) {
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: StreamResult -> OutputStream

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+    DocumentBuilder builder;
+    try {
+      builder = factory.newDocumentBuilder();
+
+      Document doc = builder.newDocument();
+      Element rootElement = doc.createElementNS("xml_steam", "XML_STEAM");
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: stream?

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+      Element objectPull = doc.createElement("Object_pull");
+      Element objectSteam = doc.createElement("Object_steam");
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: ObjectStream?

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  public void createXMLDocument(Stream<Object> inputStream, StreamResult stream) {
+    DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+    DocumentBuilder builder;
+    try {
+      builder = factory.newDocumentBuilder();
+
+      Document doc = builder.newDocument();
+      Element rootElement = doc.createElementNS("xml_steam", "XML_STEAM");
+
+      Element objectPull = doc.createElement("Object_pull");
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: objectPool?

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 1 above)
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: Stream is a bad choice. It's better to create your own class like stream with methods like `add` and `flush`

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  private void parseObject(
+      Document doc,
+      Object obj,
+      Element pull,
+      Element steam,
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: stream

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+    do {
+      obj = queue.remove();
+      String className = obj.getClass().toString().substring(6);
+      className = className.replaceAll("\\$", "--subclass--");
+      Element currObject = doc.createElement(className);
+      if (parsedObjects.containsKey(obj)) {
+        currObject.setAttribute("id", String.valueOf(parsedObjects.get(obj).getKey()));
+      } else {
+        currObject.setAttribute("id", String.valueOf(id));
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: Create a method for unique id's

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+import javax.xml.transform.*;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+
+public class XMLSerializator {
+  private int id = 0;
+  private static final List<String> primAndWrappers =
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: Change to Set (`contains` for hashset is O(1) vs O(n) for list)

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+      Element objectPull = doc.createElement("Object_pull");
+      Element objectSteam = doc.createElement("Object_steam");
+      doc.appendChild(rootElement);
+      rootElement.appendChild(objectSteam);
+      rootElement.appendChild(objectPull);
+
+      TransformerFactory transformerFactory = TransformerFactory.newInstance();
+      Transformer transformer = transformerFactory.newTransformer();
+      Map<Object, Map.Entry<Integer, String>> parsedObjects = new IdentityHashMap<>();
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: <Object, Integer>?

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+            }
+          }
+
+          Element elem = doc.createElement(name);
+          elem.setAttribute("type", type.toString());
+          elem.appendChild(doc.createTextNode(value));
+          currObject.appendChild(elem);
+
+        } catch (IllegalAccessException e) {
+          e.printStackTrace();
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: Process exception

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+      inputStream.forEach(x -> parseObject(doc, x, objectPull, objectSteam, parsedObjects));
+
+      transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+      DOMSource source = new DOMSource(doc);
+
+      transformer.transform(source, stream);
+
+    } catch (Exception e) {
+      e.printStackTrace();
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: Process exception

**Review by SpykeX3 — CHANGES_REQUESTED** [review]

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 6 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: it's unique -_-

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 8 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: yes, it's absolutely useless string, fixed

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 7 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: fixed

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 1 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: fixed

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 4 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: fixed

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 3 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: fixed

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 2 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: fixed

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 1 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: fixed

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 11]
```diff
@@ -15,96 +16,118 @@
...
+              "java.lang.Boolean",
+              "java.lang.Character",
+              "java.lang.String"));
+
+  public void write(Object obj) throws NullPointerException {
+    if (obj == null) throw new NullPointerException();
+    queue.add(obj);
+  }
+
+  public void flush(OutputStream stream) {
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: Move to constructor

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    try {
+      builder = factory.newDocumentBuilder();
+    } catch (ParserConfigurationException e) {
+      throw new RuntimeException(e);
+    }
+
+    Document doc = builder.newDocument();
+    Element rootElement = doc.createElementNS("xml_stream", "XML_STEAM");
+
+    Element objectPull = doc.createElement("Object_pool");
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: pool

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+    Element objectSteam = doc.createElement("Object_stream");
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: stream

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+    DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+    DocumentBuilder builder;
+    try {
+      builder = factory.newDocumentBuilder();
+    } catch (ParserConfigurationException e) {
+      throw new RuntimeException(e);
+    }
+
+    Document doc = builder.newDocument();
+    Element rootElement = doc.createElementNS("xml_stream", "XML_STEAM");
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: XML_STREAM

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+    transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+    DOMSource source = new DOMSource(doc);
+
+    try {
+      transformer.transform(source, streamResult);
+    } catch (TransformerException e) {
+      throw new RuntimeException(e);
+    }
+
+    queue = new ArrayDeque<>();
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: queue.clear()

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    parsedObjects = new HashMap<>();
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: parsedObjects.clear()

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+    id = 0;
+
+  }
+
+  private void parseObject(Document doc, Object obj, Element pull, Element stream) {
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: pull?

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+
+    if (parsedObjects.containsKey(obj)) {
+      stream.appendChild(createLink(obj, doc));
+      return;
+    }
+
+    Queue<Object> queue = new ArrayDeque<>();
+    queue.add(obj);
+    boolean isSteam = true;
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]: isSteam?

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+      parsedObjects.put(obj, parsedId);
+
+      currElement.setAttribute("id", parsedId);
+      currElement.setAttribute("type", objType.toString());
+
+      if (objType.isArray()) {
+
+        if (!primitives.contains(objType.getComponentType().toString())) {
+          Object[] array = (Object[]) obj;
+          String[] idInArray = new String[array.length];
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java:128]: Use stream and getOrDefault

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+          } catch (IllegalAccessException ignored) {
+          }
+
+          //System.out.println(name);
+          if (name.contains("$")) continue;
+          Element elem = doc.createElement(name);
+
+         // System.out.println(type);
+          elem.setAttribute("type", type.toString());
+          elem.appendChild(doc.createTextNode(value));
```
Comment (SpykeX3) [inline lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java]:
```
          try {
            value = parseObjectValue(field, obj, queue);
          } catch (IllegalAccessException ignored) {
          }
```
value may be null -> potential NPE

**Review by SpykeX3 — CHANGES_REQUESTED** [review]

File: lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java (hunk 19 above)
Comment (villanellex2) [reply lib/src/main/java/ru/nsu/fit/XMLSerialization/XMLSerializator.java:128]: нет

**Pull request merged**